File: ios/P3Fitness/ViewModels/ImportViewModel.swift
import Foundation
import SwiftData
import Observation

// MARK: - Supporting Types

enum ImportMode: String, CaseIterable {
    case url  = "URL"
    case text = "Text"
}

enum ImportState {
    case idle
    case loading
    case preview(ImportedRoutine)
    case error(String)
    case success
}

// MARK: - ImportViewModel

@Observable
final class ImportViewModel {
    var inputText: String = ""
    var importMode: ImportMode = .url
    var importState: ImportState = .idle
    var routineName: String = ""

    // MARK: - Import from URL

    @MainActor
    func importFromURL(_ urlString: String, userId: Int, context: ModelContext) async {
        importState = .loading
        do {
            let apiKeyManager = APIKeyManager.shared
            let routine = try await RoutineImporter.importFromURL(urlString, apiKeyManager: apiKeyManager)
            importState = .preview(routine)
        } catch {
            importState = .error(error.localizedDescription)
        }
    }

    // MARK: - Import from Text

    @MainActor
    func importFromText(_ text: String, userId: Int, context: ModelContext) async {
        importState = .loading
        do {
            let apiKeyManager = APIKeyManager.shared
            let routine = try await RoutineImporter.importFromText(text, apiKeyManager: apiKeyManager)
            importState = .preview(routine)
        } catch {
            importState = .error(error.localizedDescription)
        }
    }

    // MARK: - Save Imported Routine

    @MainActor
    func saveImportedRoutine(routine: ImportedRoutine, customName: String, userId: Int, context: ModelContext) async {
        var finalRoutine = routine
        let trimmedName = customName.trimmingCharacters(in: .whitespacesAndNewlines)
        if !trimmedName.isEmpty {
            finalRoutine.name = trimmedName
        }

        let workoutRoutine = WorkoutRoutine(userId: userId, name: finalRoutine.name)
        context.insert(workoutRoutine)

        // Fetch all existing exercises for name-matching
        let allExDescriptor = FetchDescriptor<Exercise>()
        let allExercises = (try? context.fetch(allExDescriptor)) ?? []

        for (index, importedEx) in finalRoutine.exercises.enumerated() {
            let matchedExercise = allExercises.first {
                $0.name.lowercased() == importedEx.name.lowercased()
            }

            let exerciseId: Int
            let exerciseName: String

            if let existing = matchedExercise {
                exerciseId = existing.id
                exerciseName = existing.name
            } else {
                // Create a new custom exercise
                let newId = abs(Int(Date().timeIntervalSince1970 * 1000) % 1_000_000_000) + index + 1000
                let newEx = Exercise(
                    id: newId,
                    name: importedEx.name,
                    muscleGroup: validatedMuscleGroup(importedEx.muscleGroup),
                    equipment: validatedEquipment(importedEx.equipment),
                    isCustom: true
                )
                context.insert(newEx)
                exerciseId = newId
                exerciseName = importedEx.name
            }

            let re = RoutineExercise(
                routineId: workoutRoutine.id,
                exerciseId: exerciseId,
                exerciseName: exerciseName,
                orderIndex: index,
                defaultSets: max(1, importedEx.sets),
                defaultReps: max(1, importedEx.reps)
            )
            context.insert(re)
        }

        do {
            try context.save()
            importState = .success
        } catch {
            importState = .error("Failed to save routine: \(error.localizedDescription)")
        }
    }

    // MARK: - Reset

    func reset() {
        inputText = ""
        importState = .idle
        routineName = ""
    }

    // MARK: - Helpers

    private func validatedMuscleGroup(_ group: String) -> String {
        Exercise.muscleGroups.first { $0.lowercased() == group.lowercased() } ?? "Other"
    }

    private func validatedEquipment(_ equipment: String) -> String {
        Exercise.equipmentTypes.first { $0.lowercased() == equipment.lowercased() } ?? "Bodyweight"
    }
}

File: app/src/main/java/com/fittrack/ui/viewmodel/ImportViewModel.kt
package com.fittrack.ui.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.fittrack.FitTrackApplication
import com.fittrack.data.api.ApiKeyManager
import com.fittrack.data.api.ImportState
import com.fittrack.data.api.ImportedRoutine
import com.fittrack.data.api.RoutineImporter
import com.fittrack.data.entity.Exercise
import com.fittrack.data.entity.RoutineExercise
import com.fittrack.data.entity.WorkoutRoutine
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

data class ImportUiState(
    val mode: ImportMode = ImportMode.URL,
    val urlInput: String = "",
    val textInput: String = "",
    val importState: ImportState = ImportState.Idle,
    val editedRoutineName: String = "",
    val hasApiKey: Boolean = false
)

enum class ImportMode { URL, TEXT }

class ImportViewModel(
    application: Application,
    private val userId: Int
) : AndroidViewModel(application) {
    private val db = (application as FitTrackApplication).database
    private val routineDao = db.workoutRoutineDao()
    private val routineExerciseDao = db.routineExerciseDao()
    private val exerciseDao = db.exerciseDao()
    private val apiKeyManager = ApiKeyManager(application)
    private val importer = RoutineImporter(apiKeyManager)

    private val _uiState = MutableStateFlow(ImportUiState(hasApiKey = apiKeyManager.hasApiKey()))
    val uiState: StateFlow<ImportUiState> = _uiState.asStateFlow()

    fun setMode(mode: ImportMode) {
        _uiState.value = _uiState.value.copy(mode = mode, importState = ImportState.Idle)
    }

    fun updateUrl(url: String) {
        _uiState.value = _uiState.value.copy(urlInput = url)
    }

    fun updateText(text: String) {
        _uiState.value = _uiState.value.copy(textInput = text)
    }

    fun updateRoutineName(name: String) {
        _uiState.value = _uiState.value.copy(editedRoutineName = name)
    }

    fun startImport() {
        val state = _uiState.value
        viewModelScope.launch(Dispatchers.IO) {
            when (state.mode) {
                ImportMode.URL -> importer.importFromUrl(state.urlInput) { result ->
                    _uiState.value = _uiState.value.copy(
                        importState = result,
                        editedRoutineName = if (result is ImportState.Preview) result.routine.name else ""
                    )
                }
                ImportMode.TEXT -> importer.importFromText(state.textInput) { result ->
                    _uiState.value = _uiState.value.copy(
                        importState = result,
                        editedRoutineName = if (result is ImportState.Preview) result.routine.name else ""
                    )
                }
            }
        }
    }

    fun saveImportedRoutine(routine: ImportedRoutine) {
        viewModelScope.launch(Dispatchers.IO) {
            val name = _uiState.value.editedRoutineName.ifBlank { routine.name }
            val routineId = routineDao.insert(
                WorkoutRoutine(userId = userId, name = name)
            ).toInt()

            val allExercises = exerciseDao.getAllExercises().first()

            routine.exercises.forEachIndexed { index, imported ->
                // Find best match from existing exercises
                val match = allExercises.firstOrNull {
                    it.name.equals(imported.name, ignoreCase = true)
                }

                val exerciseId = match?.id ?: exerciseDao.insert(
                    Exercise(
                        name = imported.name,
                        muscleGroup = imported.muscleGroup.ifBlank { "Other" },
                        equipment = imported.equipment.ifBlank { "Bodyweight" },
                        isCustom = true
                    )
                ).toInt()

                routineExerciseDao.insert(
                    RoutineExercise(
                        routineId = routineId,
                        exerciseId = exerciseId,
                        orderIndex = index,
                        defaultSets = imported.sets,
                        defaultReps = imported.reps
                    )
                )
            }

            _uiState.value = _uiState.value.copy(importState = ImportState.Success)
        }
    }

    fun resetState() {
        _uiState.value = ImportUiState(hasApiKey = apiKeyManager.hasApiKey())
    }

    class Factory(
        private val application: Application,
        private val userId: Int
    ) : ViewModelProvider.Factory {
        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return ImportViewModel(application, userId) as T
        }
    }
}
